[Merged by Bors] - Add Distance and Atmospheric Fog support

coreh opened the pull request:
<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/203873533-44c029af-13b7-4740-8ea3-af96bd5867c9.png">
<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/203873549-36be7a23-b341-42a2-8a9f-ceea8ac7a2b8.png">


# Objective

- Add support for the “classic” distance fog effect, as well as a more advanced atmospheric fog effect.

## Solution

This PR:

- Introduces a new `FogSettings` component that controls distance fog per-camera. 
- Adds support for three widely used “traditional” fog falloff modes: `Linear`, `Exponential` and `ExponentialSquared`, as well as a more advanced `Atmospheric` fog;
- Adds support for directional light influence over fog color;
- Extracts fog via `ExtractComponent`, then uses a prepare system that sets up a new dynamic uniform struct (`Fog`), similar to other mesh view types;
- Renders fog in PBR material shader, as a final adjustment to the `output_color`, after PBR is computed (but before tone mapping);
- Adds a new `StandardMaterial` flag to enable fog; (`fog_enabled`)
- Adds convenience methods for easier artistic control when creating non-linear fog types;
- Adds documentation around fog.

---

## Changelog

### Added

- Added support for distance-based fog effects for PBR materials, controllable per-camera via the new `FogSettings` component;
- Added `FogFalloff` enum for selecting between three widely used “traditional” fog falloff modes: `Linear`, `Exponential` and `ExponentialSquared`, as well as a more advanced `Atmospheric` fog;

alice-i-cecile commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,171 @@
+    /// <path d="M263 151V49" stroke="#FF00E5" stroke-dasharray="1 4"/>
+    /// <text font-family="sans-serif" fill="#FF00E5" style="white-space: pre" font-family="Inter" font-size="10" letter-spacing="0em"><tspan x="121" y="58.6364">start</tspan></text>
+    /// <text font-family="sans-serif" fill="#FF00E5" style="white-space: pre" font-family="Inter" font-size="10" letter-spacing="0em"><tspan x="267" y="58.6364">end</tspan></text>
+    /// </svg>
```
TIL you can embed SVG graphs inside of rustdoc!

coreh replied:
I saw the `Color` constant docs that use styled `<span>`s and decided to give it a try, was also pleasantly surprised

alice-i-cecile commented on crates/bevy_pbr/src/render/fog.rs:
```diff
@@ -0,0 +1,70 @@
+use crate::{Fog, FogMode};
+
+#[derive(Copy, Clone, ShaderType, Default, Debug)]
+pub struct GpuFog {
```
Basic doc strings :)

alice-i-cecile reviewed (commented):
Very cool, and the code and docs look good (caveat that I'm still learning rendering). I do wonder if we can make these more... modular? Having to include this sort of effect in bevy_pbr, and adding a bitflag to `StandardMaterial` definitely feels like it might get messy.

alice-i-cecile commented on crates/bevy_pbr/src/render/fog.wgsl:
```diff
@@ -0,0 +1,29 @@
+    distance: f32,
+) -> vec4<f32> {
+    var result = fog.color;
+    result.a *= 1.0 - 1.0 / exp(pow(distance * fog.density_or_start, 2.0));
```
Is using pow 2.0 here really going to be competitive with another two multiplications? Do wgsl compilers optimize this for you?

coreh replied:
No idea... If the WGSL doesn't do it, maybe the underlying GLSL/SPIRV/MSL compiler does?

Is there a good way to benchmark this? It probably adds up but hard to isolate in the middle of all of the other stuff the shader is doing

alice-i-cecile replied:
Hmm, unsure 🤔 Maybe try stress testing it on a very busy scene like animated foxes and then check the traces? Either way, this was mostly curiosity.

coreh commented:
Thanks for the speedy CR! Added additional documentation comments.

> I do wonder if we can make these more... modular? Having to include this sort of effect in bevy_pbr, and adding a bitflag to StandardMaterial definitely feels like it might get messy.

Not sure, there are already _some_ things that affect for example lights that end up as bitflags there instead of special light-related components, so it's not completely out of place, but maybe there's indeed a cleaner way? I think what `Wireframe` is doing is maybe too heavy handed for this, but I'm new to the bevy way of doing things. Is there some precedent in how to modularize stuff like this?

I also got some feedback in `#rendering-dev` on Discord about maybe moving `Fog` itself from a resource to a component, so that it can very per-camera, not sure how to implement extraction for that case.

alice-i-cecile commented:
> Maybe? There are already some things that affect, for example, lights that end up as bitflags there instead of special light-related components, so it's not completely out of place, but maybe there's indeed a cleaner way?

Yep, not a problem for this PR, but wanted to surface the idea so we start chewing on it.

> I also got some feedback also in #rendering-dev about maybe moving Fog itself from a resource to a component, so that it can vary per-camera, not sure how to implement extraction for that case.

Oh I rather like that. IIRC there's an [`ExtractComponentPlugin`](https://dev-docs.bevyengine.org/bevy/render/extract_component/struct.ExtractComponentPlugin.html) that should point you in the right direction.

coreh commented:
> IIRC there's an [ExtractComponentPlugin](https://dev-docs.bevyengine.org/bevy/render/extract_component/struct.ExtractComponentPlugin.html) that should point you in the right direction.

Thanks! That was a good starting point.

`Fog` is now a `Component` and can be set independently per-camera. Here's a modified `split_screen` example running with two different fogs, one for each camera:

<img width="1392" alt="Screenshot 2022-11-01 at 02 19 21" src="https://user-images.githubusercontent.com/418473/199166040-6f869945-3bd5-46cd-bea3-a8f0b03b7da9.png">

I also consolidated the fog setup code into a single `FogPlugin`.

---

Some ergonomics/usability questions:

1. Now that we can add/remove fog from cameras by simply adding and removing the component, does it still make sense to have a `FogMode::Off` variant? Is there some precedent to Components that are “added but disabled”, and does that feel idiomatic?
2. If we decide to keep the `Off` variant, should it still be the default? Or should we make one of the other variants a more “useful” out-of-the-box default?
3. Should we keep the short name `Fog` or make it something more specific (e.g. `DistanceFog`)? That's longer to type/remember but could makes sense in a future where there are other fog effects which are _not_ added to the camera, but rather as separate, independent entities with `Transform`/`GlobalTransform` (e.g. `LinearFog`, `SphericalFog` or even `VolumetricFog`), kinda like lights work currently.

mockersf commented:
Could you add an example that show off the various fog modes? Maybe cycle through them on a timer, or on a key press

JMS55 commented:
1. I would remove the Off variant, and make the component not-added by default.
2. I would go with the more verbose option, something like DistanceFogSettings (and DistanceFogPlugin).

In general, this is the patterns I've been seeing for post-processing effects recently:
* Make an ExamplePlugin plugin, that's either loaded by default as part of PbrPlugin (Bloom), or optionally added by the user (FXAAPlugin)
* Make an ExampleSettings component to be applied to any view you want the effect on, whose presence enables the effect and which holds the settings (no off switch on this component).
* Optionally, add a setting to PbrPlugin or somewhere similar to globally toggle the effect off.

coreh commented:
Added an interactive example, where you can tweak the parameters:

https://user-images.githubusercontent.com/418473/199402028-8d651f3f-bf33-4afa-bd89-28d251880fdb.mov

alice-i-cecile commented on examples/3d/fog.rs:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    if keycode.pressed(KeyCode::LBracket) {
+        let r = (fog.color.g() - 0.1 * delta).max(0.0);
```
You should swap your variable names for the latter if statements to be less confusing :p

coreh replied:
Oops.

alice-i-cecile commented on examples/3d/fog.rs:
```diff
@@ -0,0 +1,310 @@
+        .push_str(format!("\n\n1 / 2 / 3 - Switch Fog Mode").as_str());
+
+    if keycode.pressed(KeyCode::Key1) {
+        fog.mode = match fog.mode {
```
Nit: I think these statements are likely clearer with `if let` branches.

alice-i-cecile commented on examples/3d/fog.rs:
```diff
@@ -0,0 +1,310 @@
+//! This example shows how to use distance fog
```
Describing the controls here would be nice. Full keybindings in a single little table would be great.

coreh commented:
> I would go with the more verbose option, something like DistanceFogSettings (and DistanceFogPlugin).

Renamed `Fog` to `FogSettings`. I'm going for this shorter name instead of `DistanceFogSettings` because I believe we can reuse these same settings for volumetric fogs, so in the future you could add `FogCubeVolume { .. }` and `FogSettings { .. }` to a non-camera entity to produce volumetric fog.

Added the atmospheric extinction/inscattering falloff calculations (in a new `Atmospheric` fog falloff mode) and the pseudo-sun scattering math from [Inigo Quilez's article](https://iquilezles.org/articles/fog/), as suggested by @superdump:

<img width="1392" alt="Screenshot 2022-11-03 at 03 27 10" src="https://user-images.githubusercontent.com/418473/199661317-6468c326-1681-4e82-a608-21fdbf2828f6.png">
<img width="1392" alt="Screenshot 2022-11-03 at 03 27 06" src="https://user-images.githubusercontent.com/418473/199661338-f5b2c064-8c61-489e-93b8-930921817697.png">
<img width="1392" alt="Screenshot 2022-11-03 at 03 10 52" src="https://user-images.githubusercontent.com/418473/199661372-fbdf39bc-ce83-4f43-8ab4-1a7a81ccc680.png">
<img width="1392" alt="Screenshot 2022-11-03 at 03 15 29" src="https://user-images.githubusercontent.com/418473/199661410-85defbbd-57e4-4c9e-814d-4968e4a0804f.png">

coreh commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,223 @@
+    /// Produces a “halo” or light dispersion effect (e.g. around the sun)
+    pub scattering_color: Color,
+
+    /// The expoent applied to the directional light alignment calculation
+    /// A higher value means a more concentrated “halo”
+    pub scattering_expoent: f32,
```
Oof, looks like I feel for the _slightly_ differently-spelled cognate 😅 I didn't realize there was an extra “n” there in English

<img width="666" alt="image" src="https://user-images.githubusercontent.com/418473/199871065-2a3c85d1-4586-40ed-839b-ebdc4bfa9c11.png">

coreh commented on crates/bevy_pbr/src/render/fog.wgsl:
```diff
@@ -0,0 +1,67 @@
+        result = vec4<f32>(mix(result.rgb, fog.scattering_color.rgb, min(scattering * fog.scattering_color.a, 1.0)), result.a);
+    }
+
+    let extinction_color = 1.0 - 1.0 / vec3<f32>(exp(distance * fog.be.r), exp(distance * fog.be.g), exp(distance * fog.be.b));
+    let inscattering_color = 1.0 - 1.0 / vec3<f32>(exp(distance * fog.bi.r), exp(distance * fog.bi.g), exp(distance * fog.bi.b));
```
Unfortunately I don't think this will work, since $k \cdot e^{x} \ne e^{k \cdot x}$.

superdump replied:
The wgsl `*` of two vectors is component-wise multiplication and exp() on a vector is also component-wise. So changing vec3(exp(v.x), exp(v.y), exp(v.z))

coreh replied:
Oh, I think you accidentally had the parenthesis in the wrong spot. In that case you mean:

```suggestion
    let extinction_color = 1.0 - 1.0 / exp(vec3<f32>(distance * fog.be));
    let inscattering_color = 1.0 - 1.0 / exp(vec3<f32>(distance * fog.bi));
```

That should indeed work. I didn't think of using per-component `exp()`. Let me tweak the code to use it.

superdump replied:
No. The parentheses are in the right position as far as I can see. `distance` is a scalar. `vec3<f32>(distance)` makes it into a 3D vector with the same value for each component. This vector value can be reused for both `be` and `bi` computations. And `*` multiplies vector component-wise in wgsl and other shader languages, it isn't a dot/scalar product.

coreh replied:
Ouch, I misread GitHub's diff because of the highlighting it applies, I read it as if the pairing was between `exp(` and the highlighted parameter `)` 🤦. You're 100% right.

<img width="284" alt="image" src="https://user-images.githubusercontent.com/418473/199882993-e509696b-0801-4c44-968b-166b8b4d2952.png">

coreh replied:
I did some experiments and it looks like it's okay with splatting the entire thing implicitly, so this works, which is even shorter without the explicit `vec3<f32>`:

```
    let extinction_color = 1.0 - 1.0 / exp(distance * fog.be);
    let inscattering_color = 1.0 - 1.0 / exp(distance * fog.bi);
```

coreh commented on crates/bevy_pbr/src/pbr_material.rs:
```diff
@@ -204,6 +204,9 @@ pub struct StandardMaterial {
     pub unlit: bool,
 
+    /// Whether to disable fog for this material.
+    pub no_fog: bool,
```
We can, I used a negative form mainly to make it consistent with `unlit`, and also because I thought maybe novice users would be confused with setting `fog_enabled: true` and not getting any fog.

BTW, do you think a per-entity setting would make sense, like `NotShadowCaster`, i.e. `NotFogAffected` instead of, _or in addition to_, a material flag?

superdump replied:
I think setting no_fog: false is more confusing. :)

A per-mesh entity flag could be good but it could be done in a follow up PR if you prefer.

IceSentry replied:
> `fog_enabled: true` and not getting any fog.

Couldn't you just invert it were you are using it?

coreh replied:
Changed it, it's now a positive rather than a negative flag. Yeah, I think I prefer to do the entity flag on a follow up PR, this branch/thread is already running quite long, it's getting somewhat hard to follow.

coreh commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -6,6 +6,7 @@
 #import bevy_pbr::clustered_forward
 #import bevy_pbr::lighting
 #import bevy_pbr::shadows
+#import bevy_pbr::fog
```
I added it here for consistency with the other ones like `shadows` and `lighting`, which if I understand correctly are also only used by `pbr_functions.wgsl`... I don't fully understand `#import` behavior (and in fact much of) WGSL so I'm mostly cargo culting things 😅 as I go along

superdump replied:
This is fine.

coreh commented:
> Also, any chance of adding a demo of the final mode to the example? :)

I'm thinking about maybe adding a separate demo, since the scene on the `fog` demo doesn't have sufficiently apart far/near objects for the different color channel-specific effect of extinction/in-scattering to be really noticeable.

coreh commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,223 @@
+
+    /// The color used for the fog the view direction aligns with directional lights
+    /// Produces a “halo” or light dispersion effect (e.g. around the sun)
+    pub scattering_color: Color,
```
> I think this variable name could be improved. The best I can come up with right now is `directional_light_color`.

👍 Renamed it to that, also renamed `scattering_exponent` to `directional_light_exponent` for consistency.

> That makes me think that now we support multiple directional lights, we could apply this for each light source and use the light's colour instead of setting it here... :)

Modified it so that it takes the light color into account, but also multiplies it with this value, to preserve some degree of artistic control. (e.g. in case you want the atmosphere glow form the sun to look purple-ish but don't want your directional light to be _actually_ purple)

coreh commented:
Added an example showcasing the atmospheric fog effect, with toggles for disabling it, and also disabling the directional light influence.

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/199911136-5c1d55c0-e57c-4e94-b192-6c25ebf3cc07.png">
<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/199911225-976621ef-513b-4534-b4a2-babe26e89113.png">
<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/199911256-43c1f2b7-656b-490e-9d87-bce3f71912ef.png">

JMS55 commented on examples/3d/fog.rs:
```diff
@@ -0,0 +1,326 @@
+
+fn setup_instructions(mut commands: Commands, asset_server: Res<AssetServer>) {
+    // UI camera
+    commands.spawn(Camera2dBundle {
```
Same here

superdump commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,223 @@
+    /// <text font-family="sans-serif" fill="currentColor" style="white-space: pre" font-size="12" letter-spacing="0em"><tspan x="161" y="190.864">distance</tspan></text>
+    /// <text font-family="sans-serif" transform="translate(10 132) rotate(-90)" fill="currentColor" style="white-space: pre" font-size="12" letter-spacing="0em"><tspan x="0" y="11.8636">fog intensity</tspan></text>
+    /// </svg>
+    Exponential { density: f32 },
```
This was resolved but was there an answer?

coreh commented on examples/3d/atmospheric_fog.rs:
```diff
@@ -0,0 +1,125 @@
+
+fn setup_instructions(mut commands: Commands, asset_server: Res<AssetServer>) {
+    // UI camera
+    commands.spawn(Camera2dBundle {
```
Nice 🥳

coreh commented:
Did some reading about atmospheric scattering and visibility, and I now have a much better grasp of what the parameters do and how to control them.

Added some utility methods based on Koschmieder's visibility equation: 

 - For `FogFalloff::Exponential`:
     - `FogFalloff::from_visibility()`
     - `FogFalloff::from_visibility_contrast()`

 - For `FogFalloff::ExponentialSquared`:
     - `FogFalloff::from_visibility_squared()`
     - `FogFalloff::from_visibility_contrast_squared()`

 - For `FogFalloff::Atmospheric`:
     - `FogFalloff::from_visibility_color()`
     - `FogFalloff::from_visibility_colors()`
     - `FogFalloff::from_visibility_contrast_color()`
     - `FogFalloff::from_visibility_contrast_colors()`

For atmospheric fog, these also convert colors into the right exponential scale needed for the `extinction` and `inscattering` parameters, so it's now much easier to have artistic control over the look of the fog.

cart commented:
I've removed this from the 0.9 milestone. Sadly its just too big to properly review this late in the game.

coreh commented:
That's completely understandable, no worries! 😉 I plan on using `main` for the foreseeable future for my project. Please take your time

coreh commented:
BTW, merged `main`, and confirmed that @aevyrie's PR also fixes the banding effect on the fog:

### Before

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/201450593-c0f98737-d94f-4016-b799-85d4e4c212c8.png">

### After

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/418473/201450533-d6b79ad5-357b-447d-a9a0-f110c903381a.png">

aevyrie commented:
FYI, this should reduce excess noise in darker areas of banded fog, when not using HDR: https://github.com/bevyengine/bevy/pull/6707

superdump commented on crates/bevy_pbr/src/render/fog.wgsl:
```diff
@@ -0,0 +1,66 @@
+    let extinction_factor = 1.0 - 1.0 / exp(distance * fog.be);
+    let inscattering_factor = 1.0 - 1.0 / exp(distance * fog.bi);
+
+    return vec4<f32>(input_color.rgb * (1.0 - extinction_factor * fog_color.a) + fog_color.rgb * inscattering_factor * fog_color.a, input_color.a);
```
```suggestion
    return vec4<f32>(
        input_color.rgb * (1.0 - extinction_factor * fog_color.a)
            + fog_color.rgb * inscattering_factor * fog_color.a,
        input_color.a
    );
```

superdump commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+    /// <text font-family="sans-serif" transform="translate(10 132) rotate(-90)" fill="currentColor" style="white-space: pre" font-size="12" letter-spacing="0em"><tspan x="0" y="11.8636">fog intensity</tspan></text>
+    /// </svg>
+    Exponential {
+        /// Unitless multiplier applied to the world distance (within the exponential fog falloff calculation).
```
Is it really unitless? Density suggests some measure of mass or matter per volume. Do you just mean it is not physically-based?

coreh replied:
You're right, it's _not_ unitless, I think I wrote this before I had a good understanding of what the math was actually doing, and missed it when revising the docs. If I understand it correctly, the unit should be $m^{-1}$. It's also _kinda_ physically based, less so than the atmospheric fog, but still an approximation/simplification. I'll just remove the word `Unitless`.

superdump commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+    /// <text font-family="sans-serif" transform="translate(10 132) rotate(-90)" fill="currentColor" style="white-space: pre" font-size="12" letter-spacing="0em"><tspan x="0" y="11.8636">fog intensity</tspan></text>
+    /// </svg>
+    ExponentialSquared {
+        /// Unitless multiplier applied to the world distance (within the exponential squared fog falloff calculation).
```
Same here - unitless or just not physically-based?

superdump commented on crates/bevy_pbr/src/render/pbr_functions.wgsl:
```diff
@@ -268,3 +268,32 @@ fn dither(color: vec4<f32>, pos: vec2<f32>) -> vec4<f32> {
+        let n_directional_lights = lights.n_directional_lights;
+        for (var i: u32 = 0u; i < n_directional_lights; i = i + 1u) {
+            let light = lights.directional_lights[i];
+            scattering += pow(max(dot(view_to_world_normalized, light.direction_to_light), 0.0), fog.directional_light_exponent) * light.color.rgb;
```
```suggestion
            scattering += pow(
                max(
                    dot(view_to_world_normalized, light.direction_to_light),
                    0.0
                ),
                fog.directional_light_exponent
            ) * light.color.rgb;
```

superdump reviewed (commented):
This is looking basically ready to merge, just some minor wording and formatting comments.

JMS55 commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+use bevy_render::{color::Color, extract_component::ExtractComponent, prelude::Camera};
+
+/// Configures the “classic” computer graphics [distance fog](https://en.wikipedia.org/wiki/Distance_fog) effect,
+/// in which objects appear progressively more covered in atmospheric haze as they move further away from the camera.
```
```suggestion
/// in which objects appear progressively more covered in atmospheric haze the further away they are from the camera.
```

JMS55 commented on crates/bevy_pbr/src/render/fog.rs:
```diff
@@ -0,0 +1,147 @@
+    views: Query<(Entity, Option<&FogSettings>), With<ExtractedView>>,
+) {
+    for (entity, fog) in &views {
+        let gpu_fog = if let Some(fog) = fog {
```
Instead of using an if let else and then a match, you can use a match like Some(Linear) => ..., Some(Exponential) => ..., None => GPU_FOG_MODE_OFF.

JMS55 reviewed (approved):
Generally looks good to me. Only other thing I would suggest is to increase the font size of labels in the examples - they're too small and hard to read.

coreh commented:
Hey 👋 Sorry for taking a while to reply; I didn't have a lot of bandwidth for open source with holiday/new years. Catching up now!

coreh commented on crates/bevy_pbr/src/render/mesh_view_bindings.wgsl:
```diff
@@ -43,15 +43,17 @@ var<uniform> cluster_offsets_and_counts: ClusterOffsetsAndCounts;
 #ifdef MULTISAMPLED
-@group(0) @binding(10)
-var depth_prepass_texture: texture_depth_multisampled_2d;
 @group(0) @binding(11)
+var depth_prepass_texture: texture_depth_multisampled_2d;
+@group(0) @binding(12)
 var normal_prepass_texture: texture_multisampled_2d<f32>;
 #else
-@group(0) @binding(10)
-var depth_prepass_texture: texture_depth_2d;
 @group(0) @binding(11)
+var depth_prepass_texture: texture_depth_2d;
+@group(0) @binding(12)
```
This is a really weird diff, what I did was actually just change the `@binding` numbers from `10`, `11` to `11`, `12`.

superdump reviewed (changes requested):
Just one minor change and I think this is good to go.

MinerSebas commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -61,6 +61,31 @@ struct Lights {
+}
+
+// Important: These must be kept in sync with `fog.rs`
+let FOG_MODE_OFF: u32                   = 0u;
+let FOG_MODE_LINEAR: u32                = 1u;
+let FOG_MODE_EXPONENTIAL: u32           = 2u;
+let FOG_MODE_EXPONENTIAL_SQUARED: u32   = 3u;
+let FOG_MODE_ATMOSPHERIC: u32           = 4u;
```
If I read #7356 correctly, these let must be changed to const once it i merged.

coreh replied:
Thanks for the heads up, will keep an eye out.

robtfm commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+    /// Creates a [`FogFalloff::ExponentialSquared`] value from the given visibility distance in world units,
+    /// using the revised Koschmieder contrast threshold, [`FogFalloff::REVISED_KOSCHMIEDER_CONTRAST_THRESHOLD`].
+    pub fn from_visibility_squared(visibility: f32) -> FogFalloff {
+        FogFalloff::from_visibility_contrast(
```
```suggestion
        FogFalloff::from_visibility_contrast_squared(
```

coreh replied:
Thanks! Nice catch, I messed up while copying and pasting these

robtfm commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+        visibility: f32,
+        contrast_threshold: f32,
+    ) -> FogFalloff {
+        FogFalloff::Exponential {
```
```suggestion
        FogFalloff::ExponentialSquared {
```

robtfm commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+        contrast_threshold: f32,
+    ) -> FogFalloff {
+        FogFalloff::Exponential {
+            density: FogFalloff::koschmieder(visibility, contrast_threshold).sqrt(),
```
```suggestion
            density: FogFalloff::koschmieder(visibility, contrast_threshold) / 2.0,
```

robtfm replied:
actually hang on, this isn't right either... i assumed exp squared was using `exp(distance*density)^2` but it's actually using `exp((distance*density)^2)`.

so then the density to recover `contrast_threshold` at distance `visibility` must be `sqrt(linear density / z)` i think?
```suggestion
            density: (FogFalloff::koschmieder(visibility, contrast_threshold) / visibility).sqrt(),
```

superdump replied:
I've double-checked the maths and indeed `(FogFalloff::koschmieder(visibility, contrast_threshold) / visibility).sqrt()` in the last suggestion from @robtfm in the previous message above is correct.

superdump replied:
Good catch btw!

coreh replied:
Oh, really nice catch! I must have botched the math along the way when isolating the density and didn't realize it

robtfm commented on crates/bevy_pbr/src/fog.rs:
```diff
@@ -0,0 +1,488 @@
+        }
+    }
+
+    /// The original 2% contrast threshold as proposed by Koschmieder, being the
+    /// minimum visual contrast at which a human observer can detect an object.
+    pub const KOSCHMIEDER_CONTRAST_THRESHOLD: f32 = 0.02;
+
+    /// A revised 5% contrast threshold, deemed more realistic for typical human observers.
```
```suggestion
    /// A 2% contrast threshold was proposed by Koschmieder, being the
    /// minimum visual contrast at which a human observer can detect an object.
    /// We use a revised 5% contrast threshold, deemed more realistic for typical human observers.    
```

robtfm commented on crates/bevy_pbr/src/render/fog.wgsl:
```diff
@@ -0,0 +1,86 @@
+    distance: f32,
+    scattering: vec3<f32>,
+) -> vec4<f32> {
+    var fog_color = fog.base_color;
+    if (fog.directional_light_color.a > 0.0) {
+        fog_color = vec4<f32>(
+            fog.base_color.rgb
+                + scattering * fog.directional_light_color.rgb * fog.directional_light_color.a,
+            fog_color.a
+        );
+    }
```
this block is repeated 4 times, can we move it up to pbr_functions (or make a function here for it)?

also i don't think we need the branch here but since it's a uniform branch maybe it's better to keep it anyway ..?

coreh replied:
Just extracted it, it does look much cleaner.

robtfm commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -95,6 +96,12 @@ fn fragment(in: FragmentInput) -> @location(0) vec4<f32> {
+    // fog
+    if (fog.mode != FOG_MODE_OFF && (material.flags & STANDARD_MATERIAL_FLAGS_FOG_ENABLED_BIT) != 0u) {
+        let view_to_world = in.world_position.xyz - view.world_position.xyz;
+        output_color = apply_fog(output_color, view_to_world);
```
we should probably pass the position components separately, in case we want to add the height-based density modifications iq talks about in future

superdump replied:
I agree, though it's also easy to refactor when we actually do that so not a blocker for this PR imo.

coreh replied:
Is `view_to_world` likely to be reused for some other effect we're interested in? That's one specific scenario in which passing it from the outside could be useful: to avoid hiding the fact that it's used by the `apply_fog()` function and having it be calculated twice.

If that seems unlikely I can calculate it within the function.

robtfm replied:
i think the cost of a single subtraction isn't really worth worrying about either way. i just want the interface so that if somebody wants height-based fog, they can replace the fog shader handle and not have to replace the whole pbr shader top level.

but as superdump says, not blocking, we can always do it later when we have better support for that kind of substitution/plugin functionality.

robtfm reviewed (changes requested):
the documentation and examples are really excellent again. 

a couple of copy/paste errors in the helpers and minor bits in the shader then LGTM

superdump commented on crates/bevy_pbr/src/render/fog.rs:
```diff
@@ -0,0 +1,147 @@
+
+use crate::{FogFalloff, FogSettings};
+
+/// The GPU-side representation of the fog configuration that's sent as a uniform to the shader
+#[derive(Copy, Clone, ShaderType, Default, Debug)]
+pub struct GpuFog {
+    /// Unsigned int representation of the active fog falloff mode
+    mode: u32,
+    /// Fog color
+    base_color: Vec4,
+    /// The color used for the fog where the view direction aligns with directional lights
+    directional_light_color: Vec4,
+    /// The exponent applied to the directional light alignment calculation
+    directional_light_exponent: f32,
+    /// Allocated differently depending on fog mode.
+    /// See `mesh_view_types.wgsl` for a detailed explanation
+    be: Vec3,
+    /// Allocated differently depending on fog mode.
+    /// See `mesh_view_types.wgsl` for a detailed explanation
+    bi: Vec3,
+}
```
Reordering for better packing of data in the uniform. It makes the order of the fields a bit less nice to read, but it could bring some minor performance benefit GPU-side perhaps.
```suggestion
/// The GPU-side representation of the fog configuration that's sent as a uniform to the shader
#[derive(Copy, Clone, ShaderType, Default, Debug)]
pub struct GpuFog {
    /// Fog color
    base_color: Vec4,
    /// The color used for the fog where the view direction aligns with directional lights
    directional_light_color: Vec4,
    /// Allocated differently depending on fog mode.
    /// See `mesh_view_types.wgsl` for a detailed explanation
    be: Vec3,
    /// The exponent applied to the directional light alignment calculation
    directional_light_exponent: f32,
    /// Allocated differently depending on fog mode.
    /// See `mesh_view_types.wgsl` for a detailed explanation
    bi: Vec3,
    /// Unsigned int representation of the active fog falloff mode
    mode: u32,
}
```

superdump commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -61,6 +61,31 @@ struct Lights {
     spot_light_shadowmap_offset: i32,
 };
 
+struct Fog {
+    mode: u32,
+    base_color: vec4<f32>,
+    directional_light_color: vec4<f32>,
+    directional_light_exponent: f32,
+    // `be` and `bi` are allocated differently depending on the fog mode
+    //
+    // For Linear Fog:
+    //     be.x = start, be.y = end
+    // For Exponential and ExponentialSquared Fog:
+    //     be.x = density
+    // For Atmospheric Fog:
+    //     be = per-channel extinction density
+    //     bi = per-channel inscattering density
+    be: vec3<f32>,
+    bi: vec3<f32>,
+}
```
Reordering for better packing. May bring a slight GPU-side performance benefit perhaps.
```suggestion
struct Fog {
    base_color: vec4<f32>,
    directional_light_color: vec4<f32>,
    // `be` and `bi` are allocated differently depending on the fog mode
    //
    // For Linear Fog:
    //     be.x = start, be.y = end
    // For Exponential and ExponentialSquared Fog:
    //     be.x = density
    // For Atmospheric Fog:
    //     be = per-channel extinction density
    //     bi = per-channel inscattering density
    be: vec3<f32>,
    directional_light_exponent: f32,
    bi: vec3<f32>,
    mode: u32,
}
```

coreh commented on crates/bevy_pbr/src/render/fog.wgsl:
```diff
@@ -0,0 +1,86 @@
+             fog_color.a
+         );
+    }
+    let density = fog.be.x;
+    fog_color.a *= 1.0 - 1.0 / exp(pow(distance * density, 2.0));
```
Can't we trust the compiler or the underlying driver to optimize this `pow()` call away? Seems like a very low hanging fruit as far as optimizations go, but we can definitely do it by hand if it's not something we can rely on

robtfm replied:
there are a lot of compilers out there ... no harm in changing it.

coreh commented:
@superdump @robtfm Applied feedback and updated to the latest `main`. Added tweaks so that the shadows still look correct, as we now have cascade shadow maps (🎉!)

superdump reviewed (approved):
bors r+
